sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala

```diff
@@ -207,29 +203,143 @@ class CrossJoinIterator(
+        leftMap.close()
+        JoinGatherer(rightMap, rightBatch)
+      case (_, _) => JoinGatherer(leftMap, leftBatch, rightMap, rightBatch)
+    }
+    if (joinGatherer.isDone) {
+      joinGatherer.close()
+      None
+    } else {
+      Some(joinGatherer)
+    }
+  }
+}
+
+class ConditionalSemiOrAntiJoinIterator(
+    joinType: JoinType,
+    builtBatch: LazySpillableColumnarBatch,
+    stream: Iterator[LazySpillableColumnarBatch],
+    targetSize: Long,
+    condition: ast.CompiledExpression,
+    spillCallback: SpillCallback,
+    joinTime: GpuMetric,
+    totalTime: GpuMetric)
+    extends NestedLoopJoinIterator(
+      joinType,
+      builtBatch,
+      stream,
+      targetSize,
+      joinTime,
+      totalTime) {
+  private[this] var compiledExpr: Option[ast.CompiledExpression] = Some(condition)
+  TaskContext.get().addTaskCompletionListener[Unit](_ => close())
+
+  override def makeGatherer(streamBatch: LazySpillableColumnarBatch): Option[JoinGatherer] = {
+    withResource(GpuColumnVector.from(streamBatch.getBatch)) { leftTable =>
+      withResource(GpuColumnVector.from(builtBatch.getBatch)) { rightTable =>
+        val map = joinType match {
+          case LeftSemi =>
+            compiledExpr.map(leftTable.conditionalLeftSemiJoinGatherMap(rightTable, _, false))
+                .getOrElse(leftTable.leftSemiJoinGatherMap(rightTable, false))
+          case LeftAnti =>
+            compiledExpr.map(leftTable.conditionalLeftAntiJoinGatherMap(rightTable, _, false))
+                .getOrElse(leftTable.leftAntiJoinGatherMap(rightTable, false))
+          case _ =>
+            throw new IllegalStateException(s"Unexpected join type $joinType")
+        }
+        withResource(map) { map =>
+          val lazyMap = LazySpillableGatherMap(map, spillCallback, "left_map")
+          val gatherer = JoinGatherer(lazyMap, streamBatch)
+          if (gatherer.isDone) {
+            gatherer.close()
+            None
+          } else {
+            Some(gatherer)
+          }
+        }
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    super.close()
+    compiledExpr.foreach(_.close())
+    compiledExpr = None
+  }
+}
+
 object GpuBroadcastNestedLoopJoinExecBase extends Arm {
-  def innerLikeJoin(
+  def nestedLoopJoin(
+      joinType: JoinType,
+      numFirstTableColumns: Int,
       builtBatch: LazySpillableColumnarBatch,
       stream: Iterator[LazySpillableColumnarBatch],
       targetSize: Long,
       buildSide: GpuBuildSide,
-      boundCondition: Option[Expression],
+      boundCondition: Option[GpuExpression],
+      spillCallback: SpillCallback,
       numOutputRows: GpuMetric,
       joinOutputRows: GpuMetric,
       numOutputBatches: GpuMetric,
       joinTime: GpuMetric,
       totalTime: GpuMetric): Iterator[ColumnarBatch] = {
-    val joinIterator =
+    val joinIterator = if (boundCondition.isEmpty) {
+      // A nested loop join with no condition directly in the join is treated like a cross join.
```

Comment:
I don't think this is true.  It looks like if there is no condition, then the bound condition always returns true.

https://github.com/apache/spark/blob/c888bad6a12b45f3eda8d898bdd90405985ee05c/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala#L109

Which means if I do something stupid like. 

```
df.join(df2, Seq(), "leftAnti").show
```

I will get a cross join as output instead of nothing, which I should have gotten.

We should either fall back to the CPU in very odd cases like this, or do what Spark does and force the conditional to be `GpuLiteral(True)`.

Reply:
Actually I tested the join types and this is true for everything except left_semi and left_anti. All of the others do end up being a cross join. So the check should probably just be for semi or anti.

Reply:
Well that ended up being a fun journey.  Not only was lacking condition expression not being handled correctly for semi/anti, the count-only handling was also incorrect because it was specific to cross/inner.  I've updated the code to handle the degenerate semi/anti cases when the condition is empty as well as handling the count-only special case for those.  I ended up refactoring the code a bit into sub-methods to keep `doExecuteColumnar` from becoming a large, unwieldy function.

I also fixed an issue where `getRowCountAndClose` was releasing the GPU semaphore but never re-acquiring it before emitting the next columnar batch.

Reply:
So the comment and code is identical to what was here before. It looks like you worked around it all at a higher level, which is fine, but it would be nice to have the comments here or on the method explain what is and is not supported just to be careful

Reply:
Sorry, I meant to update the comment as part of the refactor.  I tweaked the comments and added an assert.